lib/Dialect/AIEVec/Transforms/VectorToAIEVecConversions.cpp

```diff
@@ -2801,6 +2801,95 @@ struct LowerVectorContractionOpToAIEVecMatMulPattern
   bool matMoveToAcc;
 };
 
+// Convert a `vector.transpose` op to an `aievec.shuffle` op for AIEml.
+struct LowerVectorTransposeOpToAIEVecShuffleOpPattern
+    : OpConversionPattern<vector::TransposeOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(vector::TransposeOp transpOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto resTy = transpOp.getResultVectorType();
+    auto resShape = resTy.getShape();
+    auto elemTyBitWidth = resTy.getElementTypeBitWidth();
+    auto vBitWidth = std::accumulate(resShape.begin(), resShape.end(),
+                                     elemTyBitWidth, std::multiplies<>());
+    if (vBitWidth != 512)
+      return failure();
+
+    if (elemTyBitWidth != 8 && elemTyBitWidth != 16 && elemTyBitWidth != 32)
+      return failure();
+
+    // Verify leading dimensions are all 1.
+    for (int64_t i = 0; i < static_cast<int64_t>(resShape.size() - 2); ++i)
+      if (resShape[i] != 1)
+        return failure();
+
+    // Only permutation of the 2 innermost dimensions are supported.
+    ArrayRef<int64_t> perm = transpOp.getPermutation();
+    for (int64_t i = 0; i < static_cast<int64_t>(perm.size() - 2); ++i)
+      if (perm[i] != i)
+        return failure();
+    if (perm[perm.size() - 1] != static_cast<int64_t>(perm.size() - 2))
+      return failure();
+
+    auto id = aievec::ShuffleMode::T512_1X2_LO;
+    static const aievec::ShuffleMode T8Modes[17] = {
+        id,
+        id,
+        id,
+        id,
+        aievec::ShuffleMode::T8_16X4,
+        id,
+        id,
+        id,
+        aievec::ShuffleMode::T8_8X8,
+        id,
+        id,
+        id,
+        id,
+        id,
+        id,
+        id,
+        aievec::ShuffleMode::T8_4X16};
+    static const aievec::ShuffleMode T16Modes[17] = {
+        id,
+        id,
+        aievec::ShuffleMode::T16_16X2,
+        id,
+        aievec::ShuffleMode::T16_8X4,
+        id,
+        id,
+        id,
+        aievec::ShuffleMode::T16_4X8,
+        id,
+        id,
+        id,
+        id,
+        id,
+        id,
+        id,
+        aievec::ShuffleMode::T16_2X16};
+    auto shuffleMode = aievec::ShuffleMode::T32_4X4;
```

Comment:
The shape isn't checked to be `4x4` in this case?

Reply:
It's the only possibility left.

Reply:
The input is `vector.transpose`, so technically, it can be any weird shape. It will eventually lead to an illegal `aievec.shuffle` and error out later on by the `aievec.shuffle` verifier. That sounds reasonable to me.

Reply:
It doesn't even have to be a weird shape, `vector<2x8xi32>` or `vector<2x32xi8>` would end-up with the wrong shuffle mode. I'll have to add some extra restrictions to the allowed shapes, sadly 😩